### DISCUSSION TITLE
docs/metallb/

### DISCUSSION THREAD

**Discussion by giscus** [ricsanfre/pi-cluster#302] [Announcements]
# docs/metallb/

How to configure Metal LB as load balancer in our Raspberry Pi Kubernetes cluster. How to disable default K3s load balancer and configure Metal LB.

https://picluster.ricsanfre.com/docs/metallb/

**Comment by neu7ron2** [comment]
Awesome blog! finally someone actually documented the steps to do an on-prem load balancer with K3s!, a question that is still confusing me:
I've set HA Proxy as a load balancer on a separate server (as you mentioned), everything is working fine to distribute traffic in round robin to each of the  master nodes (this seem to work fine if I setup a deployment package with type NodePort or LoadBalancer... even though the load balance external IP is still 'pending').  What is the point of installing MetalLB (which works to display the external IP correctly) if the external HaProxy is just load balancing traffic anyway? Is there any point using MetalLB or ServiceLB if i'm just using an external HaProxy to route the traffic?
(ps: You are the first person to explain why the default K3s load balancer doesn't get a Exterrnal IP, I wish K3 documented this clearly.) Thanks

**Reply by ricsanfre** [reply]
Thanks @neu7ron2 for your comments.

**HAProxy** is a cluster external service to provide load balancing only for Kubernetes API, so several nodes running the control plane can be used. HA Proxy is not performing load balancing for any of the [Kubernetes Services](https://kubernetes.io/docs/concepts/services-networking/service/).

**MetalLB** is used as service running in the cluster to provide load balancing for Kubernetes Services of type **LoadBalancer**.

Traffic to a Kubernetes Service of type LoadBalacer is not routed through HAProxy but through MetalLB. MetalLB, in layer2 mode, is responsible advertising the service to the local network. One of the nodes of the cluster start responding to ARP request for IP assigned  by LoadBalancer, so traffic for that particular IP goes directly to the node.
